backend/generator.js

```diff
@@ -11,57 +11,112 @@ exports.executeScript = function (socket, formData, callback) {
-    subProject = formData.subProject.join(",");
-    for(i=0; i<formData.subProject.length; i++){
-        subDocpath.push('docs');
-    }
-  }
-
-  const args = [
-    "-g", gitUrl,
-    "-t", docTheme,
-    "-m", email,
-    "-d", debug,
-    "-u", uniqueId,
-    "-s", subProject,
-    "-p", subDocpath.join(","),
-    "-b", targetBranch,
-    "-l", docPath
-  ];
-
-  var process = spawn("./generate.sh", args);
-
-  socketHandler.handleLineOutput(socket, process, 'logs', 4);
-  socketHandler.handleLineError(socket, process, 'err-logs');
+  var buildId = uuidV4();
+  var subProjectNames = [];
+  subProjectUrl.forEach(function (x) {
+     subProjectNames.push(x.split("/")[4].split(".")[0]);
+  });
+  var mainProjetData = {
+    email: formData.email,
+    gitUrl: formData.gitUrl,
+    docTheme: formData.docTheme,
+    debug: formData.debug,
+    uniqueId: uniqueId,
+    targetBranch: formData.targetBranch,
+    docPath: formData.docPath,
+    buildId: buildId,
+    subProjectNames: subProjectNames
+  };
 
-  process.on('exit', function (code) {
-    console.log('child process exited with code ' + code);
-    var data = { code: code, email: email, uniqueId: uniqueId, gitUrl: gitUrl };
-    if (code === 0) {
-      socketHandler.handleSocket(socket, 'success', data);
+  data.push(mainProjetData);
+  subProjectUrl.forEach(function (x) {
+    var subProjectData =  {
+      email: formData.email,
+      gitUrl: x,
+      docTheme: formData.docTheme,
+      debug: formData.debug,
+      uniqueId: uuidV4(),
+      targetBranch: formData.targetBranch,
+      docPath: "docs",
+      isSubProject: "true",
+      buildId: buildId
+    };
+    data.push(subProjectData);
+  })
+  var generatedData = Promise.all(data.map(x => generate(socket, x)))
+  generatedData.then(function (result) {
+    var flag = true;
+    var failedData;
+    result.forEach(function (x) {
+      if (x.code !== 0) {
+        flag = false;
+        failedData = x;
+      }
+    });
+    if (flag) {
+      socketHandler.handleSocket(socket, 'success', result[0]);
+      var process = spawn(`zip`,[
+        `-r`,
+        `temp/${mainProjetData.email}/${mainProjetData.buildId}.zip`,
+        `temp/${mainProjetData.email}/${mainProjetData.buildId}_preview`]);
+      process.sted
       if (callback !== undefined) {
-        callback(null, data)
+        callback(null, result[0]);
       } else {
-        mailer.sendEmail(data);
+        mailer.sendEmail(result[0]);
       }
     } else {
-      socketHandler.handleSocket(socket, 'failure', data);
+      socketHandler.handleSocket(socket, 'failure', failedData);
       if (callback !== undefined) {
         callback({
-          message: `Process exited with code : ${code}`
-        })
+          message: `Process exited with code : ${failedData.code}`
+        });
       }
     }
   });
-  return true;
 };
+
+var generate = function (socket, data) {
+  return new Promise(function(resolve, reject) {
```

Comment:
Look for https://github.com/caolan/async It's use has been recommended in nodejs backend. Promises are mainly used in the front end